lib/common/common/src/mmap_hashmap.rs

```diff
@@ -0,0 +1,233 @@
+            bufw.write_all((v.len() as u64).as_bytes())?;
+
+            // Key
+            bufw.write_all(k.as_bytes())?;
+
+            bufw.write_all(zeroes(k.len().next_multiple_of(4) - k.len()))?;
+
+            // Values
+            for i in v.iter() {
+                bufw.write_all(i.as_bytes())?;
+            }
+        }
+        bufw.flush()?;
+
+        Ok(())
+    }
+
+    /// Load the hash map from file.
+    pub fn open(path: &Path) -> io::Result<Self> {
+        let file = File::open(path)?;
+
+        // SAFETY: Assume other processes do not modify the file.
+        // See https://docs.rs/memmap2/latest/memmap2/struct.Mmap.html#safety
+        let mmap = unsafe { Mmap::map(&file)? };
+
+        let header = Header::read_from_prefix(mmap.as_ref()).ok_or(io::ErrorKind::InvalidData)?;
+
+        let phf = Function::read(&mut Cursor::new(
+            &mmap
+                .get(size_of::<Header>()..header.buckets_pos as usize)
+                .ok_or(io::ErrorKind::InvalidData)?,
+        ))?;
+
+        Ok(MmapHashMap { mmap, header, phf })
+    }
+
+    /// Get the values associated with `key`. It will return `None` in the following cases:
+    /// - The key is not present in this hash map.
+    /// - The data is corrupted, e.g. invalid UTF-8 or wrong alignment. (should not happen during
+    ///   normal operation)
+    pub fn get(&self, key: &str) -> Option<&[u32]> {
+        let hash = self.phf.get(key)?;
+
+        let buckets = u64::slice_from(self.mmap.get(
+            self.header.buckets_pos as usize
+                ..self.header.buckets_pos as usize
+                    + self.header.buckets_count as usize * size_of::<u64>(),
+        )?)?;
+        let bucket_val = *buckets.get(hash as usize)?;
+
+        let entry = &self.mmap.get(
+            self.header.buckets_pos as usize
+                + self.header.buckets_count as usize * size_of::<u64>()
+                + bucket_val as usize..,
+        )?;
+
+        let key_len = u64::read_from(entry.get(0..size_of::<u64>())?)? as usize;
+        let values_len =
+            u64::read_from(entry.get(size_of::<u64>()..2 * size_of::<u64>())?)? as usize;
+
+        let hash_key =
+            str::from_utf8(entry.get(2 * size_of::<u64>()..2 * size_of::<u64>() + key_len)?)
+                .ok()?;
+
+        if hash_key != key {
+            return None;
+        }
+
+        u32::slice_from(entry.get(
+            2 * size_of::<u64>() + key_len.next_multiple_of(4)
+                ..2 * size_of::<u64>()
+                    + key_len.next_multiple_of(4)
+                    + values_len * size_of::<u32>(),
+        )?)
+    }
+}
+
+fn entry_size(key: &str, values: &BTreeSet<u32>) -> usize {
+    2 * size_of::<u64>()
+        + key.as_bytes().len().next_multiple_of(4)
+        + values.len() * size_of::<u32>()
+}
+
+/// Returns a reference to a slice of zeroes of length `len`.
+#[inline]
+fn zeroes(len: usize) -> &'static [u8] {
```

Comment:
Because we use this and other methods only for tests in this file, I propose to move `zeroes`, `random_map`, `rand_ident` into `mod tests` scope

Reply:
`zeroes` are used in the implementation, not tests.
As for `random_map`, `random_ident`, I have plans to use them later in the benchmarks, meaning that they should not be inside `mod tests`.

Reply:
Okay, resolved